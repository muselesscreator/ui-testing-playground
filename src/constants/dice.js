import { StrictDict } from 'utils';

export const diceTypes = StrictDict({
  d4: 'd4',
  d6: 'd6',
  d8: 'd8',
  d10: 'd10',
  d12: 'd12',
  d20: 'd20',
  d100: 'd100',
});

export const diceSides = StrictDict({
  [diceTypes.d4]: 4,
  [diceTypes.d6]: 6,
  [diceTypes.d8]: 8,
  [diceTypes.d10]: 10,
  [diceTypes.d12]: 12,
  [diceTypes.d20]: 20,
  [diceTypes.d100]: 100,
});

const tgcURI = 'https://s3.amazonaws.com/files.thegamecrafter.com';
export const diceURIs = StrictDict({
  [diceTypes.d4]: `${tgcURI}/23ab580e2280cd20f2ef36260eb2b8902132e5f7`,
  [diceTypes.d6]: `${tgcURI}/19337628523391821dd8aa251a925afb611932fe`,
  [diceTypes.d8]: `${tgcURI}/c123717767de732ae400c7cc0e193581ee444987`,
  [diceTypes.d10]: 'https://iambesser.me/DA/images/10/1D10.png',
  [diceTypes.d12]: `${tgcURI}/292a945e3ab648e1e24ffa6c8e130c68a8cfa3ff`,
  [diceTypes.d20]: 'https://i.ya-webdesign.com/images/d20-1-png-3.png',
  [diceTypes.d100]: 'https://cdn01.zipify.com/images/000/182/977/original/2179683_20171023T152757.png',
});
