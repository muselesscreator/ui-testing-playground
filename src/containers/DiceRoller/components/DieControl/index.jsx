/** @module */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Button, Image } from '@edx/paragon';

import actions from 'data/actions';
import selectors from 'data/selectors';

import {
  diceSides,
  diceURIs,
} from 'constants/dice';

import './DieControl.scss';

/**
 * Simple Dice-roll control component.
 * Provides a button for rolling the die, a visual display of the die type,
 * and a textual display of the most recent roll of this die.
 *
 * @param {string} type - die type ({ diceTypes } from 'variables/dice')
 */
export class DieControl extends React.Component {
  constructor(props) {
    super(props);
    this.roll = this.roll.bind(this);
  }

  /**
   * Generate a new value based on the die type, update lastRoll,
   * and call addRoll prop with roll string.
   */
  roll() {
    const { type } = this.props;
    const sides = diceSides[type];
    const value = Math.floor(Math.random() * sides + 1);
    this.props.addRoll({ type, value });
  }

  render() {
    const { type } = this.props;
    return (
      <div className="die-control">
        <Image
          className="die-img"
          src={diceURIs[type]}
          alt="dice"
        />
        <br />
        <div className="last-value">
          { this.props.lastRoll && `Last Roll: ${this.props.lastRoll}`}
        </div>
        <Button className="roll-btn" onClick={this.roll}>
          Roll 1{type}
        </Button>
      </div>
    );
  }
}
DieControl.defaultProps = {
  lastRoll: null,
};
DieControl.propTypes = {
  type: PropTypes.string.isRequired,
  // redux
  addRoll: PropTypes.func.isRequired,
  lastRoll: PropTypes.number,
};

export const mapStateToProps = (state, ownProps) => ({
  lastRoll: selectors.app.lastRollByType(state, { type: ownProps.type }),
});

export const mapDispatchToProps = {
  addRoll: actions.app.addRoll,
};

export default connect(mapStateToProps, mapDispatchToProps)(DieControl);
