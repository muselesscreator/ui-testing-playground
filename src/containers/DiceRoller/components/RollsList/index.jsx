/** @module */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import actions from 'data/actions';
import selectors from 'data/selectors';

import { Button } from '@edx/paragon';

import './RollsList.scss';

/**
 * Simple display component for the D&D dice rolls list.
 * Provides a formatted list of roll strings, with a button at the bottom
 * allowing the user to clear all registered rolls.
 */
export const RollsList = ({ clearRolls, rolls }) => (
  <div className="rolls-list">
    <div className="rolls-container">
      { rolls.map((roll) => (
        <div className="roll-entry" key={roll.rollIndex}>
          {roll.summary}
        </div>
      )) }
    </div>
    <div className="action-bar">
      <Button className="clear-rolls-btn" onClick={clearRolls}>
        Clear
      </Button>
    </div>
  </div>
);
RollsList.propTypes = {
  // redux
  rolls: PropTypes.arrayOf(PropTypes.shape({
    rollIndex: PropTypes.number,
    summary: PropTypes.string,
  })).isRequired,
  clearRolls: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  rolls: selectors.app.rollSummaries(state),
});

export const mapDispatchToProps = {
  clearRolls: actions.app.clearRolls,
};

export default connect(mapStateToProps, mapDispatchToProps)(RollsList);
