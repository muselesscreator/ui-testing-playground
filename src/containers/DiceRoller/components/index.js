export { default as DieControl } from './DieControl';
export { default as RollsList } from './RollsList';
