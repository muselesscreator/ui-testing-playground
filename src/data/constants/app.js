import { getConfig } from '@edx/frontend-platform';

// eslint-disable-next-line import/prefer-default-export
export const routePath = `${getConfig().PUBLIC_PATH}:courseId`;
