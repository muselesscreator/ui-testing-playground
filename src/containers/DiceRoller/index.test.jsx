import React from 'react';
import { shallow } from 'enzyme';

import { diceTypes } from 'constants/dice';
import { DiceRoller } from '.';
import { DieControl } from './components';

jest.mock('./components', () => ({
  DieControl: () => 'DieControl',
  RollsList: () => 'RollsList',
}));

let el;

describe('DiceRoller component', () => {
  test('snapshot', () => {
    expect(shallow(<DiceRoller />)).toMatchSnapshot();
  });
  describe('component', () => {
    beforeEach(() => {
      el = shallow(<DiceRoller />);
    });
    test('Die control for each diceType', () => {
      const dieControls = el.find(DieControl);
      expect(dieControls.length).toEqual(7);
      expect(dieControls.at(0).props().type).toEqual(diceTypes.d4);
      expect(dieControls.at(1).props().type).toEqual(diceTypes.d6);
      expect(dieControls.at(2).props().type).toEqual(diceTypes.d8);
      expect(dieControls.at(3).props().type).toEqual(diceTypes.d10);
      expect(dieControls.at(4).props().type).toEqual(diceTypes.d12);
      expect(dieControls.at(5).props().type).toEqual(diceTypes.d20);
      expect(dieControls.at(6).props().type).toEqual(diceTypes.d100);
    });
  });
});
