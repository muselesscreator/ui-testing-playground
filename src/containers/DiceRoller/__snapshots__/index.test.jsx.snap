// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiceRoller component snapshot 1`] = `
<div
  className="dice-roller-view"
>
  <div
    className="dice-container"
  >
    <DieControl
      type="d4"
    />
    <DieControl
      type="d6"
    />
    <DieControl
      type="d8"
    />
    <DieControl
      type="d10"
    />
    <br />
    <DieControl
      type="d12"
    />
    <DieControl
      type="d20"
    />
    <DieControl
      type="d100"
    />
  </div>
  <RollsList />
</div>
`;
